Desktop Access optimizations

zmb3 opened the pull request:
This PR includes 3 changes to improve performance:

1. Reuse the same image object when encoding PNGs. This prevents us from making a 16KB allocation for every single bitmap that comes through.
2. Introduce a new PNG frame message that has a length as part of the header. The original message had no length, which required a costly decode operation just to find the message boundary.
3. Eliminate the use of `binary.Write` when encoding the most common message types. While convenient, `binary.Write` uses reflection to detect the type of the input, which slows encoding down.

zmb3 committed:
Speed up RDP bitmap handling

Most of the savings here come from reusing the image.RGBA rather
than allocating a new 16KB slice for every bitmap we receive.

zmb3 committed:
Reuse the PNG encoder at the proxy

This re-encode step still needs to be removed entirely, but this
will help.

zmb3 committed:
Speed up the proxy's TDP connection

The proxy currently has to decode and re-encode each PNG frame before
sending it to the browser, because the PNG frame TDP message lacks
an explicit length field.

Introduce a new TDP message: "PNG Frame 2"

This message is identical, with the exception of a length field at
the beginning. This allows decoders that don't care about the actual
PNG data (like Teleport's proxy) to skip over the encoded data without
decoding it. Benchmarks show a 97% performance improvement for each
PNG frame received.

In order to support this workflow, tdp.Conn is updated to support a
"ParseOnly" mode. It returns the message in its encoded form, ready
to be written to the websocket.

Lastly, remove the need to read a byte and unread a byte. We were
effectively checking that the TDP message ID is valid twice for
each message.

zmb3 committed:
Avoid using binary.Write where possible

This call uses reflection, which is suboptimal for 2 reasons:

1. It's slower than the alternative
2. It requires more error checking, since it can return errors other
   than those returned by the underlying Write call (which won't ever
   fail for us because bytes.Buffer.Write never errors)

zmb3 committed:
Send PNG2Frame by default

ibeckermayer commented on rfd/0037-desktop-access-protocol.md:
````diff
@@ -222,4 +222,14 @@ is enabled. It is a container for a JSON-encoded MFA payload.
+#### 27 - PNG frame 2
+
+```
+| message type (2) | png_length uint32 | left uint32 | top uint32 | right uint32 | bottom uint32 | data []byte |
````
```suggestion
| message type (27) | png_length uint32 | left uint32 | top uint32 | right uint32 | bottom uint32 | data []byte |
```

ibeckermayer commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1504,3 +1395,35 @@ func decodeString(r io.Reader, maxLen uint32) (string, error) {
+
+func (r rawMsg) Encode() ([]byte, error) { return []byte(r), nil }
+
+// writeUint16 writes v to b in big endian order
+func writeUint16(b *bytes.Buffer, v uint16) {
+	b.WriteByte(byte(v >> 8))
+	b.WriteByte(byte(v))
+}
+
+// writeUint32 writes v to b in big endian order
+func writeUint32(b *bytes.Buffer, v uint32) {
+	b.WriteByte(byte(v >> 24))
+	b.WriteByte(byte(v >> 16))
+	b.WriteByte(byte(v >> 8))
+	b.WriteByte(byte(v))
+}
+
+// writeUint64 writes v to b in big endian order
+func writeUint64(b *bytes.Buffer, v uint64) {
+	b.WriteByte(byte(v >> 56))
+	b.WriteByte(byte(v >> 48))
+	b.WriteByte(byte(v >> 40))
+	b.WriteByte(byte(v >> 32))
+	b.WriteByte(byte(v >> 24))
+	b.WriteByte(byte(v >> 16))
+	b.WriteByte(byte(v >> 8))
+	b.WriteByte(byte(v))
+}
```
Are these more performant?

zmb3 replied:
Details in the commit message, but yes. `binary.Write` accepts an `interface{}` and uses reflection to determine the type (int32, uint64, etc.).  Reflection is quite a bit slower than just writing bytes directly.

In fact, Go 1.19 introduced new functions that do exactly this, but they work on `[]byte` instead of `bytes.Buffer` (see `binary.BigEndian.AppendUint16()`, for example)

ibeckermayer commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -232,6 +250,68 @@ func decodePNGFrame(in peekReader) (PNGFrame, error) {
 
+// PNG2Frame is a newer versin of PNGFrame that includes the
+// length of the PNG data.
+type PNG2Frame PNGFrame
```
Is there a particular reason to keep the `PNGFrame` around now that we're no longer using it? Seems like we can just rename it to `PNG2Frame`, and delete all the `PNGFrame` related code.

zmb3 replied:
It may still exist in old session recordings.

ibeckermayer replied:
Oh, right.

zmb3 committed:
Fix new message ID

Co-authored-by: Isaiah Becker-Mayer <isaiah@goteleport.com>

Commit:
fixes broken loadBitmaps

Commit:
Gives tdp::Conn a ReadWraw() method, separate from ReadMessage (which was previously
named InputMessage())

Cosmetic changes: changes tdp::Conn's InputMessage and OutputMessage to ReadMessage
and WriteMessage respectively.

Commit:
removes rawMsg

zmb3 committed:
Remove unused decodeFixedSizeMessage

ibeckermayer reviewed (approved):
I added https://github.com/gravitational/teleport/issues/16999 to track the additional implementations we can add to `readRaw()` for maximal efficiency.

Also looks like CodeQL is complaining about something that may be fixable by using `strconv. ParseUint`

ibeckermayer commented:
I think we want to backport this to v9/v10 as well

zmb3 committed:
Update webassets

zmb3 committed:
Enforce maximum RDP screen size

While RDP will tolerate larger values and simply clamp them to the max,
we prefer to be more strict.

ibeckermayer commented on lib/web/desktop.go:
```diff
@@ -98,6 +103,12 @@ func (h *Handler) createDesktopConnection(
 		return trace.BadParameter("height missing or invalid")
 	}
 
+	if width > maxRDPScreenWidth || height > maxRDPScreenHeight {
+		return trace.BadParameter("screen size of %d x %d is greater than the maximum allowed by RDP (%d x %d)",
+			width, height, maxRDPScreenWidth, maxRDPScreenHeight,
+		)
+	}
```
Erroring here causes the UI to hang at the spinner, though that's a more general problem with the error handling all along this codepath. Possibly related with https://github.com/gravitational/teleport/issues/10595

zmb3 replied:
That's super annoying, but not new. Let's get a fix for that out separately (but asap)

zmb3 committed:
Remove unused func

zmb3 committed:
Document cached img

zmb3 committed:
Update test to expect PNG2Frame

nklaassen commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -90,23 +91,46 @@ func Decode(buf []byte) (Message, error) {
 	}
-	if err := in.UnreadByte(); err != nil {
+
+	switch mt := MessageType(t); mt {
+	case TypePNG2Frame:
+		return readRawPNG2Frame(t, in)
```
I wonder if, rather than special-casing the PNG2frame here and elsewhere, it would be cleaner to have `PNG2Frame` implement a trivial `decode`/`Encode` step that just saves and returns the same byte slice. There could be a new method to actually get the png image which would do any expensive real decoding (possibly cached).

zmb3 replied:
I had something similar originally, but we ended up setting on this approach. To be honest, I don't see how that would change much. It would just move the special case to a different spot (whether you call the expensive decode operation or not).

The only reason its special cased here is because this is the only message that gets sent at high enough frequency to really matter when it comes to performance. We may eventually implement "read raw" versions for all of the other messages (they just happen to be small enough and cheap enough that decoding and re-encoding doesn't matter as much)

nklaassen commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -241,22 +333,14 @@ type MouseMove struct {
 	return buf.Bytes(), nil
 }
 
 func decodeMouseMove(in peekReader) (MouseMove, error) {
```
most of these functions no longer require a `peekReader`, just an `io.Reader` would be fine, right? Not only would it be more correct to update these to just `io.Reader`, I think this would be much easier to reason about and review this change if `peekReader` was removed or renamed, so that the compiler would enforce that all current references to `peekReader` are updated to avoid mistakenly calling `ReadByte` (now that we no longer call `UnreadByte` above)

zmb3 replied:
Changed most of them to `io.Reader`.

We could get rid of `peekReader` entirely, though it would be a little bit annoying to read a single byte via `io.Reader.Read` anywhere we need a byte at a time.

nklaassen replied:
yeah it's alright, I suggested it mostly to guarantee we didn't forget to update any `decode` functions because I'm not extremely familiar with this part of the code, but I'm happy to trust you

I think `peek` often implies you can do something like `Peek()` or `UnreadByte` where you can "peek" at something without affecting further reads, maybe changing the name to `byteReader` would kind of avoid using that and also guarantee we don't miss any `peekReader` users here? Feel free to ignore if you disagree, I will approve once the lint is fixed

zmb3 replied:
I agree with changing the name. Will do. We used to actually read and unread a byte and I removed that because it felt silly to validate that same byte and the top level and in each individual decode func.

zmb3 committed:
Remove peekReader where possible

zmb3 committed:
Merge branch 'master' into zmb3/bitmap-fun
